ttnn/cpp/ttnn/tensor/tensor.cpp

```diff
@@ -430,26 +430,20 @@ Tensor Tensor::to(Device* target_device, const MemoryConfig& mem_config) const {
 
 Tensor Tensor::to(DeviceMesh* device_mesh, const MemoryConfig& mem_config) const {
     ZoneScoped;
-    return this->to(device_mesh->get_devices(), mem_config);
+    std::vector<Device*> workers_to_use = distribute_tensor_to_mesh(*this, *device_mesh);
+    return this->to(workers_to_use, mem_config);
 }
 
 Tensor Tensor::to(const std::vector<Device*>& workers, const MemoryConfig& mem_config) const {
     ZoneScoped;
     TT_FATAL(
         validate_worker_modes(workers), "All device threads/workers must be running in the same mode (ASYNC or SYNC)");
-    // When broadcasting a single shard to all devices, we use all workers.
-    // When sending a MultiDeviceHost tensor to the cluster, send it only to devices for which shards exist
-    auto workers_to_use = workers;
-    if (std::holds_alternative<MultiDeviceStorage>(this->get_storage()) or
-        std::holds_alternative<MultiDeviceHostStorage>(this->get_storage())) {
-        workers_to_use = std::vector<Device*>(workers.begin(), workers.begin() + num_buffers_in_tensor(*this));
-    }
-    Tensor device_tensor = Tensor(workers_to_use);
+    Tensor device_tensor = Tensor(workers);
     uint32_t device_tensor_ref_count = device_tensor.tensor_attributes->record_main_thread_ref_count();
     uint32_t original_tensor_ref_count = this->tensor_attributes->record_main_thread_ref_count();
-    uint32_t num_workers = workers_to_use.size();
-    for (int worker_index = 0; worker_index < workers_to_use.size(); ++worker_index) {
-        auto& worker = workers_to_use[worker_index];
+    uint32_t num_workers = workers.size();
+    for (int worker_index = 0; worker_index < workers.size(); ++worker_index) {
+        auto& worker = workers[worker_index];
         worker->push_work([worker, *this, device_tensor, mem_config, num_workers, worker_index]() mutable {
```

Comment:
Could we run into an issue here, where the number of workers/devices we try to send a tensor to is greater than the number of SingleDevice buffers in the tensor?

Reply:
Thanks, I've added new checks to safeguard against this. This is strictly unsupported right now. We're guaranteed to have a  1:1 mapping between a shard to a worker. We can relax this in the future, if there's some motivation for doing so, but I prefer to keep it simple for now.